smart-contracts/assembly/contracts/pool.ts

```diff
@@ -0,0 +1,463 @@
+    .expect('TokenIn is missing or invalid');
+
+  const amountIn = args.nextU256().expect('AmountIn is missing or invalid');
+
+  const tokenAAddress = bytesToString(Storage.get(tokenAAddressKey));
+  const tokenBAddress = bytesToString(Storage.get(tokenBAddressKey));
+
+  // check if the token address is one of the two tokens in the pool
+  assert(
+    tokenInAddress == tokenAAddress || tokenInAddress == tokenBAddress,
+    'Invalid token address',
+  );
+
+  // Calculate fees
+  const feeRate = _getFeeRate(); // e.g., 3 => 3%
+  const feeShareProtocol = _getFeeShareProtocol(); // e.g., 50 => 50%
+
+  // totalFee = (amountIn * feeRate) / 100
+  const totalFee = getInputAmountNet(amountIn, feeRate);
+
+  // protocolFee = (totalFee * feeShareProtocol) / 100
+  const protocolFee = getInputAmountNet(totalFee, feeShareProtocol);
+
+  // lpFee = totalFee - protocolFee
+  const lpFee = u256.sub(totalFee, protocolFee);
+
+  // netInput = amountIn - totalFee
+  const netInput = u256.sub(amountIn, totalFee);
+
+  // get the address of the other token in the pool
+  const tokenOutAddress =
+    tokenInAddress == tokenAAddress ? tokenBAddress : tokenAAddress;
+
+  // get the reserves of the two tokens in the pool
+  const reserveIn = reserves.get(new Address(tokenInAddress), u256.Zero);
+  const reserveOut = reserves.get(new Address(tokenOutAddress), u256.Zero);
+
+  // calculate the amount of tokens to be swapped
+  const amountOut = getAmountOut(netInput, reserveIn, reserveOut);
+
+  // esnure that the amountOut is greater than zero
+  assert(amountOut > u256.Zero, 'AmountOut is less than or equal to zero');
+
+  // transfer the amountIn to the contract
+  new IMRC20(new Address(tokenInAddress)).transfer(Context.callee(), amountIn);
+
+  // transfer the amountOut to the caller
+  new IMRC20(new Address(tokenOutAddress)).transferFrom(
+    Context.callee(),
+    Context.caller(),
+    amountOut,
+  );
+
+  // Update reserves:
+  // The input reserve increases by netInput + lpFee (the portion of fees that goes to the LPs).
+  // The protocolFee is not added to reserves. Instead, we store it separately.
+  const newReserveIn = u256.add(reserveIn, u256.add(netInput, lpFee));
+  const newReserveOut = u256.sub(reserveOut, amountOut);
+
+  reserves.set(new Address(tokenInAddress), newReserveIn);
+  reserves.set(new Address(tokenOutAddress), newReserveOut);
+
+  // Accumulate protocol fees
+  if (protocolFee > u256.Zero) {
+    _addProtocolFee(tokenInAddress, protocolFee);
+  }
+
+  generateEvent(
+    `Swap: In=${amountIn.toString()} of ${tokenInAddress}, Out=${amountOut.toString()} of ${tokenOutAddress}, Fees: total=${totalFee.toString()}, protocol=${protocolFee.toString()}, lp=${lpFee.toString()}`,
+  );
+}
+
+/**
+ * Claims accumulated protocol fees for a given token.
+ * @param tokenAddress - Address of the token to claim fees for.
+ * @returns void
+ */
+export function claimProtocolFees(binaryArgs: StaticArray<u8>): void {
+  _onlyOwner();
```

Comment:
Not only owner can claim, anyone can claim because it will go to the owner of the Registry.